PubSub/Redis/RedisClient.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace BeyondCode\LaravelWebSockets\PubSub\Redis;
+
+use BeyondCode\LaravelWebSockets\PubSub\PubSubInterface;
+use BeyondCode\LaravelWebSockets\WebSockets\Channels\ChannelManager;
+use Clue\React\Block;
+use Clue\React\Redis\Client;
+use Clue\React\Redis\Factory;
+use Illuminate\Support\Str;
+use React\EventLoop\LoopInterface;
+use React\Promise\PromiseInterface;
+
+class RedisClient implements PubSubInterface
+{
+
+    const REDIS_KEY = ':websockets:replication:';
+    protected $apps;
+    protected $loop;
+    protected $serverId;
+    protected $publishClient;
+    protected $subscribeClient;
+
+    public function __construct()
+    {
+        $this->apps     = collect(config('websockets.apps'));
+        $this->serverId = Str::uuid()->toString();
+    }
+
+    public function publish(string $appId, array $payload): bool
+    {
+        $payload['appId']    = $appId;
+        $payload['serverId'] = $this->serverId;
+        $this->publishClient->publish(self::REDIS_KEY, json_encode($payload));
+        return true;
+    }
+
+    public function subscribe(LoopInterface $loop): PubSubInterface
+    {
+        $this->loop = $loop;
+        [$this->publishClient, $this->subscribeClient] = Block\awaitAll([$this->publishConnection(), $this->subscribeConnection()], $this->loop);
```

Comment:
To be honest I'm more wondering why you are blocking here. Is this bit ran inside the websocket server, or somewhere outside it in blocking PHP land?

Reply:
If you don't await, you will end up returning a promise. Which isn't okay if you want the publish client to end up in the container. So what we do here is block until the promises resolve on the loop, and then push the actual instances on the variables instead of promises. Hopefully that makes sense.

Reply:
So this is going to be slightly confusing, because I don't fully understand Ratchet. But my understanding is that there is an event loop that is always running that you can push processes onto. The loop is started from the console command here: https://github.com/beyondcode/laravel-websockets/blob/c1f6ffa51b60a896d44fff490f1f6631b1758c53/src/Console/StartWebSocketServer.php#L36

You don't need to restart the loop. Just block for a split second to get the redis connection promise to resolve, and then the loop continues.

Since this subscribe command only runs once, it will just connect when you boot up the websocket server, and then continue along in the background.

Reply:
That DI waiting client is a very interesting way to do it, and I'm totally open to it.

Reply:
Sorry for the triple comment, but one other thing to consider will be implementing the interface in another client such as zeromq or some other pub/sub framework.

Reply:
No worries I'm on the ReactPHP core team. Anyway in short: the event loop is the core of any ReactPHP application, on which Ratchet, on which this package is build. And the loop is essentially a: 

```php
while (true) {
    if ($stopped === true) {
        break;
    }
    // Do stuff
}
```

Once started it will started it will continue running until there is nothing left to do. You start it with `$loop->run()` and stop it with `$loop->stop()`. Once stopped it won't start on it's own again unless you call `$loop->run()` again later in the code. (And if you haven't reset it everything is `pauzed`. (Not advisable because connections are waiting and will time out at some point.)) Now the `await*` functions will call `$loop->stop()` once the promises resolve/reject. Hence my questions and my guess your redis connections are intiated before the server starts. But assuming (and you should) the one loop (to rule them all) is shared all across the application so everything hooks into the same loop. (Using more then one loop leads to unforeseen complications and problems.)

> That DI waiting client is a very interesting way to do it, and I'm totally open to it.

Cool I'll put it in a separate package tomorrow.

> Sorry for the triple comment, but one other thing to consider will be implementing the interface in another client such as zeromq or some other pub/sub framework.

Cool, but you probably will run into the same return type issue with promises again. Maybe this could/should be solved at the interface level and it's users.

Reply:
Thanks for the clarification on the loop. Makes a lot more sense.

>Cool, but you probably will run into the same return type issue with promises again. 

Yeah I guess this is my only question. Using the Waiting Client seems like the perfect way to hook onto it as a subscriber, but how should the publisher be passed back?

Reply:
Waiting client stand alone package is up at https://packagist.org/packages/wyrihaximus/react-redis-waiting-client I'll be adding integration tests for subscribing to that over the next few days

Reply:
Just to clarify, because it's clear the blocking approach was taken from my comment here: https://github.com/simonhamp/laravel-echo-ratchet-server/issues/2#issue-312013545, the reason I did it that way initially I think is just to lower the startup time by doing both connections at once. It's not strictly necessary. And like @snellingio said, it only ever happens right before the server is started, so there's no concern of the loop stopping.

Reply:
> Waiting client stand alone package is up at [packagist.org/packages/wyrihaximus/react-redis-waiting-client](https://packagist.org/packages/wyrihaximus/react-redis-waiting-client) I'll be adding integration tests for subscribing to that over the next few days

A better alternative for that package is up: https://github.com/clue/reactphp-redis/pull/82

Reply:
The lazy client release is out. That should make some thing simpler: https://github.com/clue/reactphp-redis/releases/tag/v2.3.0

Reply:
Thanks @WyriHaximus 

I'll try to pick up this PR in a bit, might make a new PR based on this one to finish it up.